### PR TITLE
all: remove forkchoicer and reorgNeeded

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This PR changes how we view sidechains. Before the merge it was possible to import a chain with lower td and not set it as canonical in InsertChain. After the merge, we expect every chain that we get via InsertChain to be canonical, non-canonical blocks can still be inserted with InsertBlockWIthoutSetHead.

If during the InsertChain, the existing chain is not canonical anymore, we mark it as a sidechain and send the SideChainEvents normally.

### CODE REVIEW

**Comment by rjl493456442** [comment]
core/blockchain_insert.go:174:27: func `(*insertIterator).first` is unused (unused)

Please fix the lint

File: core/blockchain_test.go [hunk 1]
```diff
@@ -2872,12 +2864,19 @@ func testSideImportPrunedBlocks(t *testing.T, scheme string) {
...
-	blockToReimport := blocks[5:8]
-	_, err = chain.InsertChain(blockToReimport)
-	if err != nil {
-		t.Errorf("Got error, %v", err)
-	}
+	// TODO (MariusVanDerWijden) I have no idea why this is supposed to not fail
+	// Imo were reimporting stuff from the canonical chain that was pruned away,
+	// so it should fail with pruned ancestor or something similar.
+	// it fails with triedb parent layer missing, which sounds similar to me.
+	// But I have no idea why it suddenly breaks after I removed the Forkchoicer
```
Comment (rjl493456442) [inline core/blockchain_test.go:2871]: Originally the re-imported blocks will not be executed with a lower td, it's changed in this pull request.

In path-based scheme, the pre-state of `blocks[5]` is not recoverable as state history is not maintained. You can apply this diff for fixing it

```diff
diff --git a/core/blockchain_test.go b/core/blockchain_test.go
index 7944aef49e..2521c04db2 100644
--- a/core/blockchain_test.go
+++ b/core/blockchain_test.go
@@ -22,6 +22,7 @@ import (
 	"math/big"
 	"math/rand"
 	"os"
+	"path"
 	"sync"
 	"testing"
 	"time"
@@ -2837,7 +2838,21 @@ func testSideImportPrunedBlocks(t *testing.T, scheme string) {
 	// Generate and import the canonical chain
 	_, blocks, _ := GenerateChainWithGenesis(genesis, engine, 2*TriesInMemory, nil)
 
-	chain, err := NewBlockChain(rawdb.NewMemoryDatabase(), DefaultCacheConfigWithScheme(scheme), genesis, nil, engine, vm.Config{}, nil)
+	// Construct a database with freezer enabled
+	datadir := t.TempDir()
+	ancient := path.Join(datadir, "ancient")
+
+	db, err := rawdb.Open(rawdb.OpenOptions{
+		Directory:         datadir,
+		AncientsDirectory: ancient,
+		Ephemeral:         true,
+	})
+	if err != nil {
+		t.Fatalf("Failed to create persistent database: %v", err)
+	}
+	defer db.Close()
+
+	chain, err := NewBlockChain(db, DefaultCacheConfigWithScheme(scheme), genesis, nil, engine, vm.Config{}, nil)
 	if err != nil {
 		t.Fatalf("failed to create tester chain: %v", err)
 	}
@@ -2869,14 +2884,12 @@ func testSideImportPrunedBlocks(t *testing.T, scheme string) {
 	// so it should fail with pruned ancestor or something similar.
 	// it fails with triedb parent layer missing, which sounds similar to me.
 	// But I have no idea why it suddenly breaks after I removed the Forkchoicer
-	/*
-		// Now re-import some old blocks
-		blockToReimport := blocks[5:8]
-		_, err = chain.InsertChain(blockToReimport)
-		if err != nil {
-			t.Errorf("Got error, %v", err)
-		}
-	*/
+	// Now re-import some old blocks
+	blockToReimport := blocks[5:8]
+	_, err = chain.InsertChain(blockToReimport)
+	if err != nil {
+		t.Errorf("Got error, %v", err)
+	}
 }
 
 // TestDeleteCreateRevert tests a weird state transition corner case that we hit

```

**Review by rjl493456442 — APPROVED** [review]

**Review by holiman — APPROVED** [review]
Change touches a lot of places, but it's pretty straight-forward and simple changes. LGTM

**Comment by islishude** [comment]
> core/blockchain_test.go too many arguments in call to NewBlockChain\n\thave (ethdb.Database, nil, *Genesis, nil, consensus.Engine, vm.Config, nil, nil)\n\twant (ethdb.Database, *CacheConfig, *Genesis, *ChainOverrides, consensus.Engine, vm.Config, *uint64)

It brings an error.
